### PR TITLE
[FLINK-32810] Improve memory allocation in ListStateWithCache

### PR DESCRIPTION
**Pull request by Fanoid**
## What is the purpose of the change

Right now, by default, an instance of `ListStateWithCache` uses up all the operator-scope managed memory.

It could bring some bugs in some situations, for example, when there exist more than one `ListStateWithCache`s in a single operator, or there are other places using operator-scope managed memory.

This PR tries to fix bugs with such default behaviors.

## Brief change log

- Adds `OperatorScopeManagedMemoryManager` to manage all usages of operator-scope managed memory in one operator. Every usage must register itself to the manager, then obtain the sub-fraction of memory it can use.
- Update the constructor of `ListStateWithCache` to get sub-fraction of operator-scope managed memory from an instance of `OperatorScopeManagedMemoryManager`.
- Adds unit tests for validation of the changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java [hunk 1]
```diff
@@ -62,14 +63,32 @@ public class ListStateWithCache<T> implements ListState<T> {
...
             StateInitializationContext stateInitializationContext,
             OperatorID operatorID)
             throws IOException {
+        this(
+                serializer,
+                containingTask,
+                runtimeContext,
+                stateInitializationContext,
+                operatorID,
+                0.);
```
Comment (jiangxin369) [inline flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java:79]: With this constructor, the current algorithms using `ListStateWithCache` like `Kmeans` would not use managed memory at all, which causes performance regression.

Would it be better if we divide the memory evenly if multiple `ListStateWithCache` is declared by default?

Comment (jiangxin369) [inline flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java:79]: With this constructor, the current algorithms using `ListStateWithCache` like `Kmeans` would not use managed memory at all, which causes performance regression.

Would it be better if we divide the memory evenly if multiple `ListStateWithCache` is used by default?

File: flink-ml-core/src/test/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCacheTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+public class ListStateWithCacheTest {
+
+    @Rule public TemporaryFolder tempFolder = new TemporaryFolder();
+
+    private MiniClusterConfiguration createMiniClusterConfiguration() {
+        Configuration configuration = new Configuration();
+        // Set managed memory size to a small value, so when the instance of ListStateWithCache
+        // tries to allocate memory, it will use up all managed memory assigned to itself.
+        configuration.set(TaskManagerOptions.MANAGED_MEMORY_SIZE, MemorySize.ofMebiBytes(16));
+        configuration.set(RestOptions.PORT, 18082);
```
Comment (jiangxin369) [inline flink-ml-core/src/test/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCacheTest.java:62]: Do we need this config?

**Review by jiangxin369 — COMMENTED** [review]
Thanks for the PR, left some comments.

File: flink-ml-core/src/test/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCacheTest.java (hunk 2 above)
Comment (Fanoid) [reply flink-ml-core/src/test/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCacheTest.java]: I've removed this config. Thanks.

File: flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java (hunk 1 above)
Comment (zhipeng93) [reply flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java:79]: Would it be better to set the default value as 1.0?

Comment (Fanoid) [reply flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java]: @jiangxin369 and @zhipeng93 , thanks for your comments. 

After offline discussion, I found it is necessary to do check about sub-fractions from all usages on the operator-scope managed memory. So I changed the constructor of `ListStateWithCache`.

Please check the latest changes.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/KMeans.java [hunk 3]
```diff
@@ -239,9 +240,13 @@ public void initializeState(StateInitializationContext context) throws Exception
                     context.getOperatorStateStore()
                             .getListState(new ListStateDescriptor<>("centroids", type));
 
+            OperatorScopeManagedMemoryManager manager = new OperatorScopeManagedMemoryManager();
+            manager.register("points-state", 1.);
             points =
                     new ListStateWithCache<>(
                             new VectorWithNormSerializer(),
+                            manager,
+                            "points-state",
```
Comment (jiangxin369) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/KMeans.java:249]: It is recommended to declare the reused Strings as a static final constant.

**Review by jiangxin369 — COMMENTED** [review]
@Fanoid Thanks for the updates, left some comments.

File: flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+import org.apache.flink.util.Preconditions;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A manager for operator-scope managed memory.
+ *
+ * <p>Every operator must create one (and only one) instance of this class to manage usages of
```
[the remark this answers is not in the document]
Comment (Fanoid) [reply flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java]: Good point! I've updated the changes by adding `getOrCreate` in the `OperatorScopeManagedMemoryManager`. Please check it.

**Comment by Fanoid** [comment]
hi, @jiangxin369 , would you take another look of latest changes?

File: flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java [hunk 5]
```diff
@@ -18,33 +18,75 @@
...
+            Collections.synchronizedMap(new WeakHashMap<>());
+
+    /** Stores keys and weights of memory usages. */
     protected Map<String, Double> weights = new HashMap<>();
+    /** Indicates whether the `weights` is frozen. */
     protected boolean frozen = false;
+    /** Stores sum of weights of all usages. */
     protected double sum;
 
+    OperatorScopeManagedMemoryManager() {}
```
Comment (jiangxin369) [inline flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java:58]: Can we remove this empty constructor?

Comment (jiangxin369) [inline flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java:58]: Can we remove this empty constructor?

File: flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java [hunk 6]
```diff
@@ -95,13 +92,18 @@ public ListStateWithCache(
...
-            segmentPool =
-                    new LazyMemorySegmentPool(
-                            containingTask,
-                            memoryManager,
-                            memoryManager.computeNumberOfPages(fraction * memorySubFraction));
+        if (fraction > 0) {
+            OperatorScopeManagedMemoryManager manager =
+                    OperatorScopeManagedMemoryManager.getOrCreate(operatorID);
+            double memorySubFraction = manager.getFraction(key);
+            if (fraction * memorySubFraction > 0) {
```
Comment (jiangxin369) [inline flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/ListStateWithCache.java:99]: `if (memorySubFraction > 0)` is enough.

**Review by jiangxin369 — COMMENTED** [review]
@Fanoid Thanks for the updates. LGTM overall just left two minor comments.

File: flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java (hunk 5 above)
Comment (Fanoid) [reply flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/OperatorScopeManagedMemoryManager.java]: The empty constructor is actually used to limit its access. 
But I think using `private` scope is more reasonable here. 
Please check the changes.

**Comment by Fanoid** [comment]
@jiangxin369 Thanks. Please check the changes and above responses.

**Comment by jiangxin369** [comment]
LGTM.

**Comment by zhipeng93** [comment]
Thanks @Fanoid for the fix and @jiangxin369 for the review. LGTM.

**Pull request merged**